Linking errors when compiling static smt-switch with multiple SMT solvers that share dependencies

Compiling `smt-switch` with `--static` and both Boolector and Cvc5 results in numerous linking errors of the form:
```
multiple definition of `CaDiCaL::Solver::XXXXXXXXX'
```
Where `XXXXXXXXX` can be a number of things.

This is because both solvers use Cadical under the hood (at least in our configuration) and we "repack" static libraries to include all dependencies. I'm assuming this is a well-intentioned feature that is supposed to allow users to just link against `libsmt-switch.a` and `libsmt-switch-solver.a` for the solver that they want to use. However, this breaks a whole host of assumptions that the Linux/Unix tooling ecosystem makes about how static libraries work and results in the errors above.

I think that there are two possible solutions to this:
1. We do what every other "normal" Linux/Unix static library does: produce static library archives that only contain our own object files, as well as some information (e.g. as a pkg-config `.pc` file or a CMake config file, or both) about how they ought to be linked against.
2. We preserve the current behavior where users don't need to include direct and transitive solver dependencies among the linking flags but produce a single, merged static library file that has exactly one copy of each dependency, even when multiple dependencies have the same transitive dependency, like in this case. This, however, puts a significant maintenance burden on us as we need to manage the dependencies' dependencies all on our own and need to find versions that work with each other.

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -13,7 +13,7 @@ set(INCLUDE_DIRS
   "${PROJECT_SOURCE_DIR}/include")
 
 # add to search path for find_package
-list(APPEND CMAKE_PREFIX_PATH "${PROJECT_SOURCE_DIR}/deps")
+list(APPEND CMAKE_PREFIX_PATH "${PROJECT_SOURCE_DIR}/deps/install")
 
 if (NOT SMT_SWITCH_LIB_TYPE)
   set(SMT_SWITCH_LIB_TYPE SHARED)

--- a/btor/CMakeLists.txt
+++ b/btor/CMakeLists.txt
@@ -15,7 +15,7 @@ target_include_directories (smt-switch-btor PUBLIC "${CMAKE_CURRENT_SOURCE_DIR}/
 target_include_directories (smt-switch-btor PUBLIC "${BTOR_HOME}/src")
 
 target_link_libraries(smt-switch-btor "${BTOR_HOME}/build/lib/libboolector.a")
-target_link_libraries(smt-switch-btor "${BTOR_HOME}/deps/cadical/build/libcadical.a")
+target_link_libraries(smt-switch-btor "${PROJECT_SOURCE_DIR}/deps/install/lib/libcadical.a")
 target_link_libraries(smt-switch-btor "${BTOR_HOME}/deps/btor2tools/build/lib/libbtor2parser.a")
 target_link_libraries(smt-switch-btor smt-switch)
 target_link_libraries(smt-switch-btor pthread)
@@ -29,7 +29,7 @@ if (SMT_SWITCH_LIB_TYPE STREQUAL STATIC)
     COMMAND
       mkdir smt-switch-btor && cd smt-switch-btor && ar -x "../$<TARGET_FILE_NAME:smt-switch-btor>" && cd ../ &&
       mkdir boolector && cd boolector && ar -x "${BTOR_HOME}/build/lib/libboolector.a" &&
-      ar -x "${BTOR_HOME}/deps/cadical/build/libcadical.a" &&
+      ar -x "${PROJECT_SOURCE_DIR}/deps/install/lib/libcadical.a" &&
       ar -x "${BTOR_HOME}/deps/btor2tools/build/lib/libbtor2parser.a" && cd ../ &&
       ar -qc "$<TARGET_FILE_NAME:smt-switch-btor>" ./boolector/*.o ./smt-switch-btor/*.o &&
       # now clean up

--- a/ci-scripts/setup-cadical.sh
+++ b/ci-scripts/setup-cadical.sh
@@ -1,0 +1,1 @@
+../contrib/setup-cadical.sh

--- a/contrib/setup-btor.sh
+++ b/contrib/setup-btor.sh
@@ -15,6 +15,7 @@ else
     NUM_CORES=1
 fi
 
+$DIR/setup-cadical.sh
 
 if [ ! -d "$DEPS/boolector" ]; then
     cd $DEPS
@@ -23,8 +24,7 @@ if [ ! -d "$DEPS/boolector" ]; then
     cd boolector
     git checkout -f $BTOR_VERSION
     CFLAGS="" ./contrib/setup-btor2tools.sh
-    ./contrib/setup-cadical.sh
-    ./configure.sh --only-cadical -fPIC
+    ./configure.sh --only-cadical -fPIC --path "$DEPS/install"
     cd build
     make -j$NUM_CORES
     cd $DIR
@@ -32,7 +32,7 @@ else
     echo "$DEPS/boolector already exists. If you want to rebuild, please remove it manually."
 fi
 
-if [ -f $DEPS/boolector/build/lib/libboolector.a ] && [ -f $DEPS/boolector/deps/cadical/build/libcadical.a ] && [ -f $DEPS/boolector/deps/btor2tools/build/lib/libbtor2parser.a ] ; then \
+if [ -f $DEPS/boolector/build/lib/libboolector.a ] && [ -f $DEPS/install/lib/libcadical.a ] && [ -f $DEPS/boolector/deps/btor2tools/build/lib/libbtor2parser.a ] ; then \
     echo "It appears boolector was setup successfully into $DEPS/boolector."
     echo "You may now install it with make ./configure.sh --btor && cd build && make"
 else

--- a/contrib/setup-cadical.sh
+++ b/contrib/setup-cadical.sh
@@ -1,0 +1,26 @@
+#!/bin/bash
+set -o errexit
+set -o pipefail
+set -o nounset
+
+CADICAL_VERSION=rel-1.7.4
+
+SCRIPT_NAME="$(basename "${BASH_SOURCE[0]}")"
+SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
+ROOT_DIR="$(cd "$SCRIPT_DIR/.." && pwd)"
+DEPS_DIR="$ROOT_DIR/deps"
+
+mkdir -p "$DEPS_DIR"
+cd $DEPS_DIR
+if [[ ! -d "cadical" ]]; then
+    git clone https://github.com/arminbiere/cadical
+fi
+cd cadical
+git checkout $CADICAL_VERSION
+if [[ -d "build" ]]; then
+    echo "$SCRIPT_NAME: $DEPS_DIR/cadical/build exists, skipping configure step"
+else
+    CXXFLAGS=-fPIC ./configure
+fi
+make
+install -Dm644 -t "$DEPS_DIR/install/lib" "build/libcadical.a"

--- a/contrib/setup-cadical.sh
+++ b/contrib/setup-cadical.sh
@@ -23,4 +23,5 @@ else
     CXXFLAGS=-fPIC ./configure
 fi
 make
-install -Dm644 -t "$DEPS_DIR/install/lib" "build/libcadical.a"
+mkdir -p "$DEPS_DIR/install/lib"
+install -m644 "build/libcadical.a" "$DEPS_DIR/install/lib"

--- a/contrib/setup-cadical.sh
+++ b/contrib/setup-cadical.sh
@@ -10,11 +10,14 @@ SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
 ROOT_DIR="$(cd "$SCRIPT_DIR/.." && pwd)"
 DEPS_DIR="$ROOT_DIR/deps"
 
+# Download
 mkdir -p "$DEPS_DIR"
 cd $DEPS_DIR
 if [[ ! -d "cadical" ]]; then
     git clone https://github.com/arminbiere/cadical
 fi
+
+# Build
 cd cadical
 git checkout $CADICAL_VERSION
 if [[ -d "build" ]]; then
@@ -23,5 +26,10 @@ else
     CXXFLAGS=-fPIC ./configure
 fi
 make
+
+# Install
 mkdir -p "$DEPS_DIR/install/lib"
 install -m644 "build/libcadical.a" "$DEPS_DIR/install/lib"
+mkdir -p "$DEPS_DIR/install/include"
+install -m644 "src/ccadical.h" "$DEPS_DIR/install/include"
+install -m644 "src/cadical.hpp" "$DEPS_DIR/install/include"

--- a/contrib/setup-cvc5.sh
+++ b/contrib/setup-cvc5.sh
@@ -16,13 +16,15 @@ else
     NUM_CORES=1
 fi
 
+$DIR/setup-cadical.sh
+
 if [ ! -d "$DEPS/cvc5" ]; then
     cd $DEPS
     git clone https://github.com/cvc5/cvc5.git
     chmod -R 777 cvc5
     cd cvc5
     git checkout -f ${CVC5_VERSION}
-    CXXFLAGS=-fPIC CFLAGS=-fPIC ./configure.sh --static --auto-download
+    CXXFLAGS=-fPIC CFLAGS=-fPIC ./configure.sh --static --auto-download --dep-path="$DEPS/install"
     cd build
     make -j$NUM_CORES
     cd $DIR
@@ -30,7 +32,7 @@ else
     echo "$DEPS/cvc5 already exists. If you want to rebuild, please remove it manually."
 fi
 
-if [ -f $DEPS/cvc5/build/src/libcvc5.a ] && [ -f $DEPS/cvc5/build/src/parser/libcvc5parser.a ]; then
+if [ -f $DEPS/cvc5/build/src/libcvc5.a ] && [ -f $DEPS/cvc5/build/src/parser/libcvc5parser.a ] && [ -f $DEPS/install/lib/libcadical.a ]; then
     echo "It appears cvc5 was setup successfully into $DEPS/cvc5."
     echo "You may now install it with make ./configure.sh --cvc5 && cd build && make"
 else

--- a/cvc5/CMakeLists.txt
+++ b/cvc5/CMakeLists.txt
@@ -32,7 +32,7 @@ find_library(POLYLIBXX REQUIRED
 
 find_library(CADICAL REQUIRED
              NAMES libcadical.a libcadical.so libcadical.dylib
-             PATHS "${CVC5_HOME}/build/deps/lib")
+             )
 
 target_link_libraries(smt-switch-cvc5 "${CVC5_HOME}/build/src/libcvc5.a")
 target_link_libraries(smt-switch-cvc5 "${CVC5_HOME}/build/src/parser/libcvc5parser.a")

